Cv can be uploaded on the candidate page

gabichelsea opened the pull request:
Trello card link:
[https://trello.com/c/zJsEOfPP](https://trello.com/c/zJsEOfPP)

Without tests
I think they should be subtasks

gabichelsea committed:
Added commons dependencies
Added file validation message keys
Added CommonsMultipartResolver configuration

gabichelsea committed:
Added Exceptions
Added FileValidationRule and mapper

gabichelsea committed:
Added FileValidator and impl
Added FileBucket

gabichelsea committed:
Added FileUploadController
Modified jsp

gabichelsea committed:
Modified schema:
Added cvPath to user's table

gabichelsea committed:
Corrected schema
CvPath added to Candidate table not to User table

gabichelsea committed:
FileUpload is not available when you have already a cv
Cucumber test fails yet

gabichelsea committed:
Fixed it

gabichelsea committed:
Added FileIsAlreadyExistValidationException and messageKey

gabichelsea committed:
Modified FileUploadController
Added FileDownloadController

gabichelsea committed:
Moved model classes to file package

gabichelsea committed:
Renamed and supplement FileUploaderController

gabichelsea committed:
Modified CandidateCreationController
Modified SingleCandidateController
Modified candidate_create.jsp

gabichelsea committed:
All CV status (not exist, in progress, already exist) is working
Using Session attribute instead of my CVStatusMonitor

gabichelsea committed:
Introduce FileUploadingProperties

gabichelsea committed:
Added FileIsMissingValidationException
Added regular expression (file format)

gabichelsea committed:
Moved fileUploadingProperties file to another package
The download link for cv contains the file name

gabichelsea committed:
Finished controllers

adamsoltesz commented on pom.xml:
```diff
@@ -471,6 +471,19 @@
+                <groupId>commons-io</groupId>
+                <artifactId>commons-io</artifactId>
+                <version>${commons.io.version}</version>
+                <scope>compile</scope>
```
Compile is the default scope, it doesn't need to be specified explicitly. Check below also.

gabichelsea replied:
.

adamsoltesz commented on web/src/main/java/com/epam/rft/atsy/web/controllers/CandidateCreationController.java:
```diff
@@ -2,38 +2,54 @@
+    ModelAndView modelAndView = new ModelAndView(VIEW_NAME);
     modelAndView.addObject(CANDIDATE_OBJECT_KEY, candidateService.getCandidate(candidateId));
 
+    String cvPath = session.getAttribute(FileUploadingProperties.SESSION_PARAM_CV_PATH) == null ?
```
There could be a method for this in FileUploadingProperties (it could be renamed to FileUploadingUtil)

adamsoltesz commented on web/src/main/java/com/epam/rft/atsy/web/controllers/CandidateCreationController.java:
```diff
@@ -2,38 +2,54 @@
+    } else if (candidateCvPath == null && cvPath != null) {
+      modelAndView.addObject(FileStatus.CV_STATUS, FileStatus.FILE_IS_IN_PROGRESS.getValue());
+    } else if (candidateCvPath != null) {
+      modelAndView.addObject(FileStatus.CV_STATUS, FileStatus.FILE_IS_ALREADY_EXIST.getValue());
```
To be grammatically okay, please rename FILE_IS_ALREADY_EXIST to FILE_ALREADY_EXISTS

gabichelsea replied:
.

adamsoltesz commented on web/src/main/java/com/epam/rft/atsy/web/controllers/FileDownloaderController.java:
```diff
@@ -0,0 +1,49 @@
+  public void downloadFile(@PathVariable("candidateId") Long candidateId,
+                           HttpServletResponse response) throws IOException {
+
+    String path = candidateService.getCVPathByCandidateId(candidateId);
```
What if the candidate does not have a file uploaded? 

This should be checked and handled gracefully.

adamsoltesz commented on web/src/main/java/com/epam/rft/atsy/web/controllers/FileDownloaderController.java:
```diff
@@ -0,0 +1,49 @@
+@Controller
+public class FileDownloaderController {
+  private static final String CONTENT_DISPOSITION = "Content-Disposition";
+  private static final String MIME_TYPE = "application/octet-stream";
```
This could be called DEFAULT_MIME_TYPE, because it acts as a default as I see.

adamsoltesz commented on web/src/main/webapp/WEB-INF/pages/candidate_create.jsp:
```diff
@@ -198,6 +204,71 @@
+        <c:if test="${empty candidateId}">
+        <form:form method="POST" action="${fileUpload}" enctype="multipart/form-data">
+
+                    <c:if test="${cv_status eq false}">
```
Formatting errors in the whole file. Maybe tabs are used mixed with spaces. Should be only spaces.

adamsoltesz commented on web/src/main/java/com/epam/rft/atsy/web/validator/FileValidatorImpl.java:
```diff
@@ -0,0 +1,66 @@
+
+  private static final long MAX_FILE_SIZE = 1024 * 1024 * 10; // 10 MB
+  // Example: Garfield_Hungry-Angry_2010-12-24.doc
+  private static final String FILE_PATTERN = "([A-Z][a-z]+[-|_])+[0-9-]+[.](doc|docx|odt|pdf)$";
```
I think this check should be separated. We should have a different message if the extension is not supported, and a different one if the filename (the part before the extension) contains invalid characters.

adamsoltesz replied:
The pattern is also too strict. I tried to upload a file with name:  Alma-fa.docx and it did not work. I think the characters that you allowed should be allowed in any order that is a valid filename.

Also, I think the error message should contain the list of allowed characters, and not an example correct filename.

adamsoltesz commented on web/src/main/webapp/WEB-INF/pages/candidate_create.jsp:
```diff
@@ -198,6 +204,71 @@
+        <c:if test="${cv_status eq true}">
+        <c:set var="candidateCvPath" value="${candidate.cvPath}" />
+        <c:set var="cvPathArray" value="${fn:split(candidateCvPath, '\\\\')}" />
+        <c:set var="cvName" value="${cvPathArray[fn:length(cvPathArray)-1]}" />
```
CV name should be added by the controller to the model. We should not do this logic in the view.

adamsoltesz commented on web/src/main/webapp/WEB-INF/pages/candidate_create.jsp:
```diff
@@ -3,6 +3,7 @@
 <%@taglib uri="http://www.springframework.org/tags" prefix="spring" %>
```
Be consistent, put prefix as first attribute, so the are aligned nicely below each other, and can be easily read.

adamsoltesz commented on web/src/main/java/com/epam/rft/atsy/web/model/file/FileStatus.java:
```diff
@@ -0,0 +1,16 @@
+@AllArgsConstructor
+public enum FileStatus {
+
+  FILE_IS_NOT_EXIST(false),
```
Should be FILE_NOT_EXISTS

adamsoltesz commented on web/src/main/java/com/epam/rft/atsy/web/model/file/FileStatus.java:
```diff
@@ -0,0 +1,16 @@
+
+  FILE_IS_NOT_EXIST(false),
+  FILE_IS_IN_PROGRESS(null),
+  FILE_IS_ALREADY_EXIST(true);
```
Should be FILE_ALREADY_EXISTS

Relevart090 commented:
There is a NullPointerException in the PasswordChangeController during the running of the unit tests.

adamsoltesz commented:
The styling of the upload box is really weird. It's under the "Cancel" "Save" buttons. It looks like not part of the form. Also the error messages appear below it, and not above, this is also unusual. 

The parts of the form are too close to each other the blue button is touching the "Browse..." button.

adamsoltesz commented on web/src/main/java/com/epam/rft/atsy/web/model/file/FileStatus.java:
```diff
@@ -0,0 +1,16 @@
+  FILE_IS_IN_PROGRESS(null),
+  FILE_IS_ALREADY_EXIST(true);
+
+  public static final String CV_STATUS = "cv_status";
```
You could just return the enum values and check them in the jsp, a boolean is not too good here and not needed I think.

Also the name should contain info about it's purpose. E.g. CV_STATUS_MODEL_ATTR_NAME

adamsoltesz commented:
I see that the file is saved to disk, but it's name does not seem to be saved in the DB.

And I cannot see the download link. Maybe because of this.

gabichelsea committed:
delete compile scope from pom.xml

gabichelsea committed:
Rename statuses in FileStatus enum

gabichelsea committed:
Replace mimeType

gabichelsea committed:
In FileStatus remove getValue() method and the property for boolean values

gabichelsea committed:
Removed HttpSession params
Removed the file upload chance from the candidate page where the candidate id is null

gabichelsea committed:
CV name should be added by the controller to the model.

gabichelsea committed:
Pattern for only extension
Pattern for only fileName

gabichelsea committed:
Error message appears above the CV

gabichelsea committed:
Refactoring FileDownloadController and added validation for it

gabichelsea committed:
Upload configuration is in file_storage.properties file

lamfalusy commented on web/src/main/java/com/epam/rft/atsy/web/configuration/WebConfiguration.java:
```diff
@@ -97,4 +98,9 @@ protected void configureHandlerExceptionResolvers(
     exceptionResolvers.add(1, new UncheckedExceptionResolver(mappingJackson2JsonView()));
   }
+
+  @Bean(name = "multipartResolver")
```
name annotation attribute is unnecessary, because the bean's name is the method name by default, in this situation the method name and the value of the name attribute are the same

lamfalusy commented on web/src/main/java/com/epam/rft/atsy/web/configuration/WebConfiguration.java:
```diff
@@ -24,14 +28,31 @@
 
+  @Value("${upload_location_cv}")
+  private String uploadLocationValue;
+  public static String UPLOAD_LOCATION;
```
In my opinion this solution with static field and with FileUploadingUtil is an unnecessary magic. Fortunately there is a simpler solution: it is possible to use @Value annotation in controller classes, so FileUploadingUtil should be eliminated and use @Value("${upload_location_cv}") in FileDownloadController and FileUploadController. There is another static field in FileUploadingUtil: CV_NAME. This field is only used in CandidateCreationController, so my suggestion: move this constant to CandidateCreationController and delete FileUploadingUtil class.

lamfalusy commented on web/src/main/java/com/epam/rft/atsy/web/validator/FileValidatorImpl.java:
```diff
@@ -1,25 +1,34 @@
+  private static final String EXTENSION_PDF = "pdf";
+  private static final Collection<String> EXTENSIONS =
+      Arrays.asList(EXTENSION_DOC, EXTENSION_DOCX, EXTENSION_ODT, EXTENSION_PDF);
+
```
redundant empty line

gabichelsea committed:
Copy uploadLocation with @Value annotation to FileUploadController and FileDownloadController
Remove FileUploadingUtil

gabichelsea committed:
Removed unnecessary static variable from WebConfiguration

gabichelsea committed:
cvPath rename to cvFileName
rename getCvPathByCandidateId to getCvFilenameById

gabichelsea committed:
Added test cases for getCvFilenameById() method

gabichelsea committed:
Added FileValidationRuleMapperTest.java

gabichelsea committed:
Added FileValidatorImplTest.java

gabichelsea committed:
Added null test case for FileValidator validate() method

gabichelsea committed:
Cleaner code

gabichelsea committed:
Removed lines from SingleCandidateController
Added these lines to candidate_create.jsp and atsy-candidate-create.js

gabichelsea committed:
Removed FileBucket
Created MultipartFileCreatorTestHelper

gabichelsea committed:
Added @BeforeClass and @AfterClass methods to FileUploadControllerTest.java

gabichelsea committed:
Added validation to FileUploadController:

When you already have a CV file, but you can try to download a cv again

gabichelsea committed:
Added fileUploadController test cases:

When fileValidator throw an Exception
When the file is already exist that you want to upload

gabichelsea committed:
Modified file size fields type : int -> long

gabichelsea committed:
Added test case:

When the candidate has a cv, but you can try to upload cv again

gabichelsea committed:
Added test case when file uploading is successful

gabichelsea committed:
Check the cvFilename when file updating will be successful

gabichelsea committed:
Refactoring FileUploadControllerTest
Refactoring FileDownloadController
Added FileDownloadControllerTest with setup methods

gabichelsea committed:
Added REDIRECT_URL test for all test method

gabichelsea committed:
Added two test case for the FileDownloadControllerTest
When cvFilename is null and cvFilename is empty

gabichelsea committed:
Added CandidateCVFileHandlerUtil class to handle the file and folder creating for cv

gabichelsea committed:
Remove getCVFilenameById() method from CandidateService

gabichelsea committed:
Finish FileUploadController tests

gabichelsea committed:
Finish FileDownloadControllerTest.java

gabichelsea committed:
Removed unused import

gabichelsea committed:
Removed unique constraint on cvFilename

gabichelsea committed:
Added test case when:
File name is available in DB, but not available in File system
Candidate service return null

lamfalusy commented on web/src/main/java/com/epam/rft/atsy/web/configuration/WebConfiguration.java:
```diff
@@ -120,7 +120,7 @@ protected void configureHandlerExceptionResolvers(
   }
 
-  @Bean(name = "multipartResolver")
+  @Bean()
```
Almost! Deleting name attribute is fine, but parentheses can be deleted too.

lamfalusy commented on web/src/test/java/com/epam/rft/atsy/web/validator/FileValidatorImplTest.java:
```diff
@@ -25,11 +25,11 @@
-  private static final int FILE_SIZE_TEN_MB_MINUS_ONE_BYTE = (1024 * 1024 * 10) - 1;
-  private static final int FILE_SIZE_TEN_MB = (1024 * 1024 * 10);
-  private static final int FILE_SIZE_TEN_MB_PLUS_ONE_BYTE = (1024 * 1024 * 10) + 1;
+  private static final long FILE_SIZE_ZERO_MB = 0;
```
I do not understand the reason of this int to long type change. It seems unnecessary for me.

lamfalusy commented on web/src/test/java/com/epam/rft/atsy/helper/MultipartFileCreatorTestHelper.java:
```diff
@@ -10,9 +10,9 @@
+  public static MockMultipartFile createMultipartFile(String originalFilename, Long fileSizeByte) {
     return new MockMultipartFile(MULTIPART_FILE_NAME, originalFilename, CONTENT_TYPE,
-        getStringBySize(fileSize).getBytes());
+        getStringBySize(fileSizeByte.intValue()).getBytes());
```
It is a very dangerous solution because long can store a wider range than an integer, so intValue method will truncate those numbers, which are bigger then the biggest integer number, without any exception, so this createMultipartFile method simply will return a wrong result in case of "enough" big long fileSizeByte parameter value. Let's give it a try, call this method with 105.000.000.000 fileSizeByte parameter value, the result will not be the expected. Anyway this is a bad practice: you use a long parameter as an integer, but why? If you want to use this parameter as an integer, then the parameter type have to be integer or if it is a requirement to support long, then do it in a right and safety way.

lamfalusy commented on web/src/main/java/com/epam/rft/atsy/web/controllers/FileDownloadController.java:
```diff
@@ -37,14 +37,13 @@
 
     String cvFilename = candidateService.getCvFilenameById(candidateId);
-    if (cvFilename != null) {
+    if (cvFilename != null && !cvFilename.isEmpty()) {
```
I would like to suggest this method: org.apache.commons.lang3.StringUtils.isNotEmpty

lamfalusy commented:
Missing unit test for CandidateCVFileHandlerUtil class.

lamfalusy commented on web/src/main/java/com/epam/rft/atsy/web/util/CandidateCVFileHandlerUtil.java:
```diff
@@ -0,0 +1,37 @@
+
+  private CandidateCVFileHandlerUtil() {}
+
+  public static File createCVFileFromFolderLocationAndCandidateDtoAndCVFilename(String folderLocation, CandidateDTO candidateDTO, String cvFilename) {
```
Be careful with static methods, it is hard to mock static methods in unit tests. In this situation simply create a bean (for example with @Component annotation) instead of create class with only static methods.

lamfalusy commented on web/src/main/java/com/epam/rft/atsy/web/controllers/FileUploadController.java:
```diff
@@ -62,14 +62,15 @@ public ModelAndView uploadFile(@PathVariable("candidateId") Long candidateId,
       CandidateDTO candidateDTO = candidateService.getCandidate(candidateId);
-      if (candidateDTO.getCvFilename() != null && !candidateDTO.getCvFilename().isEmpty()) {
+      String candidateCVFilename = candidateDTO.getCvFilename();
+      if (candidateCVFilename != null && !candidateCVFilename.isEmpty()) {
```
I would like to suggest this method: org.apache.commons.lang3.StringUtils.isNotEmpty

lamfalusy commented on web/src/test/java/com/epam/rft/atsy/web/validator/FileValidatorImplTest.java:
```diff
@@ -35,6 +35,13 @@
 
   private FileValidator fileValidator = new FileValidatorImpl();
 
+
```
Delete this empty line.

lamfalusy commented on web/src/main/java/com/epam/rft/atsy/web/controllers/CandidateCreationController.java:
```diff
@@ -2,48 +2,29 @@
+  @Autowired
   private CandidateService candidateService;
 
 
```
Delete this empty line.

lamfalusy commented on service/src/test/java/com/epam/rft/atsy/service/impl/CandidateServiceImplTest.java:
```diff
@@ -53,29 +56,34 @@
+
   private CandidateEntity dummyCandidateEntity;
   private CandidateDTO dummyCandidateDto;
+  private CandidateEntity candidateEntityWithCvFilename;
```
Private field 'candidateEntityWithCvFilename' is assigned but never accessed.

lamfalusy commented on service/src/test/java/com/epam/rft/atsy/service/impl/CandidateServiceImplTest.java:
```diff
@@ -53,29 +56,34 @@
   private CandidateEntity dummyCandidateEntity;
   private CandidateDTO dummyCandidateDto;
+  private CandidateEntity candidateEntityWithCvFilename;
+  private CandidateDTO candidateDTOWithCvFilename;
```
Private field 'candidateDTOWithCvFilename' is assigned but never accessed.

lamfalusy commented on web/src/test/java/com/epam/rft/atsy/web/controllers/FileUploadControllerTest.java:
```diff
@@ -52,4 +91,66 @@ public static void doBeforeClass() throws IOException {
+
+  @Test
+  public void uploadFileShouldThrowFileValidationExceptionWhenFileIsInWrongExtension() throws Exception {
+    // Given
```
Use Given-When-Then comments consistently.

gabichelsea committed:
Fixed :
MultipartFileCreatorTestHelper.java
WebConfiguration.java

gabichelsea committed:
Removed empty lines, unused variables

gabichelsea committed:
Changed long to int --> fileSize

gabichelsea committed:
Modified CandidateCVFileHandler util class to @Component class

gabichelsea committed:
Added CandidateCVFileHandler test cases

lamfalusy commented on web/src/main/java/com/epam/rft/atsy/web/controllers/FileDownloadController.java:
```diff
@@ -45,16 +46,16 @@ public Object downloadFile(@PathVariable("candidateId") Long candidateId, Redire
-    String cvFilename = candidateDTO.getCvFilename();
-    if (cvFilename != null && !cvFilename.isEmpty()) {
+    String candidateCVFilename = candidateDTO.getCvFilename();
+    if (candidateCVFilename != null && StringUtils.isNotEmpty(candidateCVFilename)) {
```
StringUtils.isNotEmpty method can handle null value, if the parameter is null, then a method will return false value. Check the documentation: https://commons.apache.org/proper/commons-lang/javadocs/api-3.1/org/apache/commons/lang3/StringUtils.html#isEmpty%28java.lang.CharSequence%29

lamfalusy commented on web/src/main/java/com/epam/rft/atsy/web/controllers/FileUploadController.java:
```diff
@@ -63,7 +65,7 @@ public ModelAndView uploadFile(@PathVariable("candidateId") Long candidateId,
       CandidateDTO candidateDTO = candidateService.getCandidate(candidateId);
       String candidateCVFilename = candidateDTO.getCvFilename();
-      if (candidateCVFilename != null && !candidateCVFilename.isEmpty()) {
+      if (candidateCVFilename != null && StringUtils.isNotEmpty(candidateCVFilename)) {
```
StringUtils.isNotEmpty method can handle null value, if the parameter is null, then a method will return false value. Check the documentation: https://commons.apache.org/proper/commons-lang/javadocs/api-3.1/org/apache/commons/lang3/StringUtils.html#isEmpty%28java.lang.CharSequence%29

gabichelsea committed:
Replace (cvFilename!= null && StringUtils.isNotEmpty()) conditional to StringUtils.isNotBlank()

gabichelsea committed:
Merge branch 'master' into cv-can-be-uploaded-on-the-candidate-page

# Conflicts:
#	web/pom.xml
#	web/src/main/java/com/epam/rft/atsy/web/controllers/CandidateCreationController.java
#	web/src/main/java/com/epam/rft/atsy/web/controllers/rest/SingleCandidateController.java

gabichelsea committed:
Removed Candidate Name from the folder that contains the cvFile

lamfalusy commented on web/src/main/java/com/epam/rft/atsy/web/controllers/rest/SingleCandidateController.java:
```diff
@@ -19,6 +18,7 @@
 import java.util.List;
 import java.util.Locale;
 
+
```
Delete this empty line.

lamfalusy commented on web/src/main/java/com/epam/rft/atsy/web/controllers/rest/SingleCandidateController.java:
```diff
@@ -29,11 +29,14 @@
 public class SingleCandidateController {
   private static final String COMMON_INVALID_INPUT_MESSAGE_KEY = "common.invalid.input";
 
+
```
Delete this empty line.

lamfalusy commented on web/src/main/java/com/epam/rft/atsy/web/controllers/rest/SingleCandidateController.java:
```diff
@@ -29,11 +29,14 @@
   @Autowired
   private CandidateService candidateService;
 
+
```
Delete this empty line.

lamfalusy commented on web/src/main/java/com/epam/rft/atsy/web/controllers/rest/SingleCandidateController.java:
```diff
@@ -29,11 +29,14 @@
+
   @Autowired
-  private MessageSource messageSource;
+
```
Delete this empty line.

lamfalusy commented on web/src/main/java/com/epam/rft/atsy/web/controllers/rest/SingleCandidateController.java:
```diff
@@ -51,6 +54,7 @@ public ResponseEntity saveOrUpdate(@Valid @RequestBody CandidateDTO candidateDTO
       Long candidateId = candidateService.saveOrUpdate(candidateDTO);
       return new ResponseEntity<>(Collections.singletonMap("id", candidateId), HttpStatus.OK);
 
+
```
Delete this empty line.

lamfalusy commented on web/src/main/java/com/epam/rft/atsy/web/controllers/rest/SingleCandidateController.java:
```diff
@@ -59,13 +63,15 @@ public ResponseEntity saveOrUpdate(@Valid @RequestBody CandidateDTO candidateDTO
 
     for (FieldError fieldError : fieldErrors) {
       restResponse.addField(fieldError.getField(),
+
```
Delete this empty line.

lamfalusy commented on web/pom.xml:
```diff
@@ -93,6 +93,11 @@
             <groupId>commons-fileupload</groupId>
             <artifactId>commons-fileupload</artifactId>
         </dependency>
+        <dependency>
```
mysql-connector-java dependency in web layer? Why?

gabichelsea committed:
Removed empty lines after merge
